src/txreconciliation.cpp

```diff
@@ -69,10 +100,56 @@ class TxReconciliationTracker::Impl {
         return std::make_tuple(we_initiate_recon, we_respond_recon, RECON_VERSION, m_local_recon_salt);
     }
 
+    bool RegisterPeer(NodeId peer_id, bool peer_inbound,
+        bool they_may_initiate, bool they_may_respond, uint32_t recon_version, uint64_t remote_salt)
+    {
+        // We do not support reconciliation salt/version updates. We treat an attempt to update
+        // after a successful registration as a protocol violation.
+        LOCK(m_mutex);
+        if (m_states.find(peer_id) != m_states.end()) return false;
+
+        // If the peer supports the version which is lower than our, we downgrade to the version
+        // they support. For now, this only guarantees that nodes with future reconciliation
+        // versions have the choice of reconciling with this current version. However, they also
+        // have the choice to refuse supporting reconciliations if the common version is not
+        // satisfactory (e.g. too low).
+        recon_version = std::min(recon_version, RECON_VERSION);
+        // v1 is the lowest version, so suggesting something below must be a protocol violation.
+        if (recon_version < 1) return false;
+
+        auto local_salt = m_local_salts.find(peer_id);
+
+        // This function should be called only after generating the local salt.
+        if (local_salt == m_local_salts.end()) return false;
+
+        // Must match SuggestReconciling logic.
+        bool we_may_initiate = !peer_inbound, we_may_respond = peer_inbound;
+
+        bool they_initiate = they_may_initiate && we_may_respond;
+        bool we_initiate = we_may_initiate && they_may_respond;
+        // If we ever announce we_initiate && we_may_respond, this will need tie-breaking. For now,
+        // this is mutually exclusive because both are based on the inbound flag.
+        assert(!(they_initiate && we_initiate));
+
+        // The peer set both flags to false, we treat it as a protocol violation.
+        if (!(they_initiate || we_initiate)) return false;
+
+        LogPrint(BCLog::NET, "Register peer=%d for reconciling with the following params: " /* Continued */
+            "we_initiate=%i, they_initiate=%i.\n", peer_id, we_initiate, they_initiate);
+
+        uint256 full_salt = ComputeSalt(local_salt->second, remote_salt);
+
+        assert(m_states.emplace(peer_id, ReconciliationState(full_salt.GetUint64(0),
+            full_salt.GetUint64(1), we_initiate)).second);
+        return true;
+    }
+
     void ForgetPeer(NodeId peer_id)
     {
         LOCK(m_mutex);
-        if (m_local_salts.erase(peer_id)) {
+        auto salt_erased = m_local_salts.erase(peer_id);
+        auto state_erased = m_states.erase(peer_id);
+        if (salt_erased || state_erased) {
```

Comment:
Shouldn't be a `&&`, otherwise we might have staling state for the peer ?

Reply:
This line only defines whether we print logs or not, so I don't understand how does it affect the actual behavior.
The ideas is that it logs deleting the state for both pre-registered and registered peers. 

The previous 2 lines taking care of deletion are indeed `&&`.

Reply:
> The ideas is that it logs deleting the state for both pre-registered and registered peers.

I understand. Maybe maybe the log comment could have a ternary and print "salt_erased" or "state_erased" to be more meaningful. Though that's really low-level information likely only relevant for debug/hacking.